cmds/test-dev.c

```diff
@@ -0,0 +1,311 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Test device
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+#include <devices/devs.h>
+
+#include "cmd.h"
+
+
+#define TEST_DEV_TIMEOUT_MS 500
+
+#define ERASE_BEFORE 1
+#define ERASE_AFTER  2
+
+#define BUF_SIZE 257 /* Write and read size specially not aligned to any power of 2 to possibly test more edge cases. */
+
+
+static void cmd_testDevInfo(void)
+{
+	lib_consolePuts("performs simple dev read/write test, usage:test-dev [-e erase before] [-E erase after] <dev> <addr> <size>");
+}
+
+
+static void region_hexdump(addr_t start, addr_t end, addr_t offp, u8 align)
+{
+	unsigned int col;
+	static const int cols = 16;
+	addr_t ptr, offs = start;
+
+	if (align != 0) {
+		offp = (offp / cols) * cols;
+		offs = (offs / cols) * cols;
+	}
+
+	while (offs < end) {
+		lib_printf("0x%08x | ", (u32)offp);
+
+		/* Print byte values */
+		for (col = 0; col < cols; ++col) {
+			ptr = offs + col;
+
+			if ((ptr < start) || (ptr >= end)) {
+				lib_consolePuts("   ");
+				continue;
+			}
+			lib_printf("%02x ", *(u8 *)ptr);
+		}
+		lib_consolePuts("| ");
+
+		/* Print "printable" representation */
+		for (col = 0; col < cols; ++col) {
+			ptr = offs + col;
+
+			if ((ptr < start) || (ptr >= end)) {
+				lib_consolePutc(' ');
+				continue;
+			}
+
+			if (lib_isprint(*(u8 *)ptr)) {
+				lib_printf("%c", *(u8 *)ptr);
+			}
+			else {
+				lib_consolePutc('.');
+			}
+		}
+
+		lib_consolePutc('\n');
+		offp += cols;
+		offs += cols;
+	}
+}
+
+
+static void print_hline(void)
+{
+	unsigned int col;
+	for (col = 0; col < 79; ++col) {
+		lib_consolePutc('-');
+	}
+	lib_consolePutc('\n');
+}
+
+
+static int cmd_testDevWrite(int major, int minor, addr_t addr, size_t length)
+{
+	ssize_t res;
+	size_t rsz, i, chunk;
+	u8 buf[BUF_SIZE];
+
+	for (rsz = 0; rsz < length; rsz += res) {
+		chunk = min((length - rsz), BUF_SIZE);
+		for (i = 0; i < chunk; i++) {
+			buf[i] = (i & 0xff);
+		}
+		res = devs_write(major, minor, addr + rsz, buf, chunk);
```

Comment:
Move the buffer filling out of the loop.
```suggestion
	for (i = 0; i < sizeof(buf); i++) {
		buf[i] = (i & 0xff);
	}
	for (rsz = 0; rsz < length; rsz += res) {
		chunk = min((length - rsz), BUF_SIZE);
		res = devs_write(major, minor, addr + rsz, buf, chunk);
```
in
```c
	chunk = min((length - rsz), BUF_SIZE);
```
lines uniformly use either `BUF_SIZE` or `sizeof(buf)`.

Reply:
this code change and buffer can no longer be filled outside of the loop